lightning/src/chain/package.rs

```diff
@@ -809,6 +804,20 @@ impl PackageTemplate {
 			height_original,
 		}
 	}
+
+	fn map_output_type_flags(input_solving_data: &PackageSolvingData) -> (PackageMalleability, bool) {
+		let (malleability, aggregable) = match input_solving_data {
+			PackageSolvingData::RevokedOutput(RevokedOutput { is_counterparty_balance_or_non_anchors: None, .. }) => { (PackageMalleability::Malleable, true) },
+			PackageSolvingData::RevokedOutput(RevokedOutput { opt_anchors: Some(..), .. }) => { (PackageMalleability::Malleable, false) },
+			PackageSolvingData::RevokedOutput(RevokedOutput { opt_anchors: None, .. }) => { (PackageMalleability::Malleable, true) },
+			PackageSolvingData::RevokedHTLCOutput(..) => { (PackageMalleability::Malleable, true) },
+			PackageSolvingData::CounterpartyOfferedHTLCOutput(..) => { (PackageMalleability::Malleable, true) },
+			PackageSolvingData::CounterpartyReceivedHTLCOutput(..) => { (PackageMalleability::Malleable, false) },
+			PackageSolvingData::HolderHTLCOutput(..) => { (PackageMalleability::Untractable, false) },
```

Comment:
Can you point me to which changes ? Yes it modifies intentionally some changes to avoid exposures to claim batch on anchor outputs on our side though dunno if those are the changes you’re thinking of ?

Reply:
So I agree on the first point that `HolderHTLCOutput` should be `Malleable` with anchors however on the second point to aggregate HTLCs with known preimages this can be a safety risk if a) the paired HTLC-timeout is final and can be broadcast therefore breaking the batch and b) the value of the HTLC do not weight for the average feerate weight of each input ?

Note there is no such aggregation based on preimage currently: https://github.com/lightningdevkit/rust-lightning/blob/0e8da588954abc32c2a341cedf888ee00f918183/lightning/src/chain/package.rs#L836

Are you thinking about another branch ?

Note, this is not I’m strongly opposed to aggregate preimages claims, at the minimal we should it with the safe-guards mentioned above and if as such better to introduce them in another PR.

Reply:
> Maybe we can also start splitting them up once we're within this range and we've them tried in a batch a few times without getting a confirmation?

The issue I’ve in mind is in case of a batch with a number of HTLC superior to the cltv_expiry_delta and there is network mempools congestion. If the earliest HTLC-preimage is not confirmed, the counterparty HTLC-timeout can be broadcast to break the batch. The batch construction could be malleated by our counterparty and to have a gradient HTLC to conflict at each block. The max feerate of the batch could be observed by the counterparty based on a bit of probing of our fee-bumping reserve.

I think this is quite an advanced concern though in the lack of fee-bumping reserve management, it might be better to be conservative for “v0.1” of anchor outputs.

> That seems like a separate issue, no? I think we'll always claim outputs even if it's not economically rational to do so > at the current feerates. That's something we can certainly improve on, but that can happen as a follow-up and
> doesn't need to impact our aggregation logic.

In a world of limited fee-bumping reserves, low-value HTLCs could be forced to be batch with a high-value HTLC to diminish the effective feerate of the whole.

> Well the issue is that the aggregation already exists upstream with tests, and this PR is attempting to revert that as
> an unrelated change.

100% here, it was unclear to me what was the upstream state between anchor outputs PRs. We can discuss the no-batch of HTLC-preimage transactions in its own PR (I still think we should do it!)

Reply:
>  If we're super concerned about our ability to get a tx confirmed in time because we're low on fees, the worst thing we can do is (preemptively) split the packages up into lots of transactions.

Splitting packages into a lot of transactions when we’re in low-fees is far from being the less economically rational strategy to adopt if we have HTLC-preimages with strong asymmetries in term of offered HTLC outputs because they have been inbound routed to us by an adversary. With anchors, I think we’re introducing a fee-bumping reserves oracle to an adversary that can be adapted in real-time based on the packages we’re broadcasting. I share the belief on having a lot of complex never-tested code to special-case on-time package splitting and the adequate mitigation might be at the `ChannelManager`-level, namely refusing to route HTLC when we’re low on fee-bumping reserves. In anycase, if anyone has an up-to-date threat model for anchor output channels, you’re free to share it.